pangeo_forge_recipes/recipes/reference_hdf_zarr.py

```diff
@@ -0,0 +1,202 @@
+    #  a master merge in finalise. This would maybe map to iter_chunk,
+    #  store_chunk, finalize_target. The strategy was used for NWM's 370k files.
+
+    # TODO: as written, this rcipe is specific to HDF5 files. fsspec-reference-maker
+    #  also supports TIFF and grib2 (and more coming)
+    netcdf_url: Union[str, List[str]]
+    output_url: str
+    _work_dir: str = ""
+    """temporary store for JSONs"""
+    netcdf_storage_options: dict = field(default_factory=dict)
+    inline_threshold: int = 500
+    output_storage_options: dict = field(default_factory=dict)
+    template_count: Optional[int] = 20
+    xarray_open_kwargs: dict = field(default_factory=dict)
+    xarray_concat_args: dict = field(default_factory=dict)
+
+    def iter_inputs(self) -> Iterable[Hashable]:
+        return ()
+
+    @property
+    def cache_input(self) -> Callable[[Hashable], None]:
+        return no_op
+
+    @property
+    def prepare_target(self) -> Callable[[], None]:
+        """Prepare the recipe for execution by initializing the target.
+        Attribute that returns a callable function.
+        """
+        self._work_dir = tempfile.mkdtemp()
+        return no_op
+
+    def iter_chunks(self) -> Iterable[Hashable]:
+        """Iterate over all target chunks."""
+        return fsspec.open_files(self.netcdf_url, **self.netcdf_storage_options)
+
+    @property
+    def store_chunk(self) -> Callable[[Hashable], None]:
+        """Store a chunk of data in the target.
+        """
+        return functools.partial(_one_chunk, work_dir=self._work_dir)
+
+    @property
+    def finalize_target(self) -> Callable[[], None]:
+        """Final step to finish the recipe after data has been written.
+        Attribute that returns a callable function.
+        """
+        if isinstance(self.netcdf_url, str):
+            proto = fsspec.utils.get_protocol(self.netcdf_url)
+        else:
+            proto = fsspec.utils.get_protocol(self.netcdf_url[0])
+        return functools.partial(
+            _finalize,
+            work_dir=self._work_dir,
+            out_url=self.output_url,
+            out_so=self.output_storage_options,
+            remote_protocol=proto,
+            remote_options=self.netcdf_storage_options,
+            xr_open_kwargs=self.xarray_open_kwargs,
+            xr_concat_kwargs=self.xarray_concat_args,
+            template_count=self.template_count,
+        )
+
+
+def _one_chunk(of, work_dir):
+    with of as f:
+        fn = os.path.join(work_dir, os.path.basename(f.name + ".json"))
+        h5chunks = SingleHdf5ToZarr(f, _unstrip_protocol(f.name, f.fs), inline_threshold=300)
+        with open(fn, "w") as outf:
+            json.dump(h5chunks.translate(), outf)
+
+
+def _finalize(
+    work_dir,
+    out_url,
+    out_so,
+    remote_protocol,
+    remote_options,
+    xr_open_kwargs,
+    xr_concat_kwargs,
+    template_count,
+):
+    files = [os.path.join(work_dir, f) for f in os.listdir(work_dir)]
```

Comment:
This won't work in cloud environments where we don't have a shared filesystem. Use `storage.MetadataTarget` instead.

Reply:
The latest version allows for passing around the results dictionaries, which works great for dask clusters. That was a workaround to credentials issues we had when wanting to write to cloud in this first stage.

Reply:
Unfortunately the way we have implemented Pangeo Forge is the opposite of pure functions that pass results around. Everything is accomplished through side effects! You have to use one of the storage targets to communicate data across processes.